markdown/java-jvm/exceptions.md

```diff
@@ -38,33 +37,206 @@ sketch: true
 ---
 
 
-## Motivation
-Lorem Ipsum. Starte mit H2-Level.
-...
+## Fehlerfälle in Java
 
-## Folie 2
-...
+```java
+double div(int a, int b){
+    return a/b;
+}
 
-## Folie 3
-...
+div(3.5,0);
+```
+::: notes
+**Probleme**:
+*   Programm wird abstürzen, da durch '0' geteilt wird
+:::
 
-## Folie 4
-...
+## Lösung?
+
+```java
+Double div(int a, int b){
+    if(b==0) return null;
+    return a/b;
+}
+
+Double sol = div(3.5,0);
+if(sol==null){
+    //do something
+}
+else {
+    //do something else
+}
+```
+::: notes
+**Probleme**:
+
+*   Da `double` nicht `null` sein kann, muss ein `Double` Objekt erzeugt und zurückgegeben werden
+*   Der Aufrufer muss auf `null` prüfen
+*   Es wird nicht kommuniziert warum `null` zurückgegeben wird. Was ist das Problem?
+*   Was ist, wenn `null` ein gültiger Rückgabewert sein soll?
+:::
 
-## Folie 5
-...
+## Unchecked vs Checked Exceptions
 
-## Folie 6
-...
+Checked:
+  * Sind (logische) Fehler im Code
+    * Fehlerhafte Benutzereingabe
+    * Unerreichbare API
+  * Treten zur Compilezeit auf
```

Comment:
Bist Du Dir da sicher?

Reply:
wenn ich mich nicht komplett irre, passiert doch beides zur laufzeit, oder?! alles andere sind compilerfehler, dafür habe/brauche ich keine exception.

die frage ist nur, wann ist es checked und wann unchecked? für mich wirkt die unterscheidung recht willkürlich. und es wird nicht eine checked draus, wenn du die throw-klausel hinzufügst. eher andersherum: bei checked _musst_ du eine throws-klausel dranpappen (wenn du den mist nicht fängst), bei unchecked steht es dir frei.

die große preisfrage ist: nach welchem muster ist es checked und wann ist es unchecked? kann man das irgendwie an der art der exception oder ausnahmefall festmachen?

Reply:
> die frage ist nur, wann ist es checked und wann unchecked? für mich wirkt die unterscheidung recht willkürlich. und es wird nicht eine checked draus, wenn du die throw-klausel hinzufügst.

Nach dem was ich so gelesen habe: doch. Der Gedanke scheint mir auch logisch zu sein. Wenn `throw` an einer Methode steht, dann MUSS ich sie prüfen, sonst kompiliert das Programm gar nicht. Die Exception ist also checked.  

Wenn kein `throw` an der Methode steht, dann muss ich sie auch nicht prüfen, das Programm kompiliert trotzdem.

[Hier](https://falconbyte.net/java-checked-unchecked-exceptions.php) hab ich auch gelesen, dass alle von `RunTimeException` abgeleitete Exception unchecked sind. 
Da weiß ich aber nicht so 100 %, ob ich damit übereinstimme. 

> die große preisfrage ist: nach welchem muster ist es checked und wann ist es unchecked? 

Sehe ich ähnlich. Theoretisch kann ich jede Exception checked machen, aber bei welchen sollte ich das machen. 

[Hier](https://www.baeldung.com/java-checked-unchecked-exceptions) habe ich auch zwei nette Beschreibungen gefunden 

> In general, checked exceptions represent errors outside the control of the program. For example, the constructor of [FileInputStream](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/FileInputStream.html#%3Cinit%3E(java.io.File)) throws FileNotFoundException if the input file does not exist. Java verifies checked exceptions at compile-time.

> If a program throws an unchecked exception, it reflects some error inside the program logic.
For example, if we divide a number by 0, Java will throw ArithmeticException:
Java does not verify unchecked exceptions at compile-time. Furthermore, we don't have to declare unchecked exceptions in a method with the throws keyword.

Reply:
> die frage ist nur, wann ist es checked und wann unchecked? für mich wirkt die unterscheidung recht willkürlich. und es wird nicht eine checked draus, wenn du die throw-klausel hinzufügst.

Okay bei mir ist grade ein Knoten im Kopf geplatzt.
ALLE Exceptions die `RuntimeException` erweitern sind unchecked. Du kannst ja auch Methoden aufrufen, die eine unchecked Exception thrown, ohen `try catch` zu verwenden ohne das der Compiler meckert.

Reply:
Glaub wir sind jetzt mehr oder minder auf einen Nenner. Hab mein Kommentar bearbeitet und meine alternativen Fakten durchgestrichen.

Reply:
jo

Reply:
@cagix todo: was sagt Bloch checked unchecked